File: frontend/client/src/components/aerospot/services/SunService.ts
/**
 * aerospot/services/SunService.ts — solposisjon og soltider.
 *
 * Selvstendig implementasjon av standard NOAA/astronomi-formler
 * (samme algoritme-familie som suncalc). Nøyaktighet ±0.5° — mer enn
 * godt nok for fotograferings-anbefalinger.
 */

import type { LatLng, SunTimes } from "../types";
import { deg2rad, rad2deg } from "./geo";

const DAY_MS = 86_400_000;
const J1970 = 2440588;
const J2000 = 2451545;
const OBLIQUITY = deg2rad(23.4397);

function toJulian(date: Date): number {
  return date.getTime() / DAY_MS - 0.5 + J1970;
}
function fromJulian(j: number): Date {
  return new Date((j + 0.5 - J1970) * DAY_MS);
}
function toDays(date: Date): number {
  return toJulian(date) - J2000;
}

function solarMeanAnomaly(d: number): number {
  return deg2rad(357.5291 + 0.98560028 * d);
}
function eclipticLongitude(M: number): number {
  const C =
    deg2rad(1.9148) * Math.sin(M) +
    deg2rad(0.02) * Math.sin(2 * M) +
    deg2rad(0.0003) * Math.sin(3 * M);
  const P = deg2rad(102.9372);
  return M + C + P + Math.PI;
}
function declination(L: number): number {
  return Math.asin(Math.sin(L) * Math.sin(OBLIQUITY));
}
function rightAscension(L: number): number {
  return Math.atan2(Math.sin(L) * Math.cos(OBLIQUITY), Math.cos(L));
}
function siderealTime(d: number, lw: number): number {
  return deg2rad(280.16 + 360.9856235 * d) - lw;
}

/** Solens asimut (fra nord, med klokken) og elevasjon i grader. */
export function sunPosition(date: Date, pos: LatLng): { azimuthDeg: number; elevationDeg: number } {
  const lw = deg2rad(-pos.lng);
  const phi = deg2rad(pos.lat);
  const d = toDays(date);
  const M = solarMeanAnomaly(d);
  const L = eclipticLongitude(M);
  const dec = declination(L);
  const ra = rightAscension(L);
  const H = siderealTime(d, lw) - ra;

  const elevation = Math.asin(
    Math.sin(phi) * Math.sin(dec) + Math.cos(phi) * Math.cos(dec) * Math.cos(H),
  );
  const azimuth =
    Math.atan2(Math.sin(H), Math.cos(H) * Math.sin(phi) - Math.tan(dec) * Math.cos(phi)) + Math.PI;

  return { azimuthDeg: (rad2deg(azimuth) + 360) % 360, elevationDeg: rad2deg(elevation) };
}

// Soltider — NOAA "hour angle"-metoden
const J0 = 0.0009;

function julianCycle(d: number, lw: number): number {
  return Math.round(d - J0 - lw / (2 * Math.PI));
}
function approxTransit(Ht: number, lw: number, n: number): number {
  return J0 + (Ht + lw) / (2 * Math.PI) + n;
}
function solarTransitJ(ds: number, M: number, L: number): number {
  return J2000 + ds + 0.0053 * Math.sin(M) - 0.0069 * Math.sin(2 * L);
}
function hourAngle(h: number, phi: number, dec: number): number {
  return Math.acos(
    (Math.sin(h) - Math.sin(phi) * Math.sin(dec)) / (Math.cos(phi) * Math.cos(dec)),
  );
}

/** Tidspunkt (Date) solen krysser gitt elevasjon (grader); settende (kveld) eller stigende. */
function timeAtElevation(date: Date, pos: LatLng, elevationDeg: number, rising: boolean): Date | null {
  const lw = deg2rad(-pos.lng);
  const phi = deg2rad(pos.lat);
  const d = toDays(date);
  const n = julianCycle(d, lw);
  const ds = approxTransit(0, lw, n);
  const M = solarMeanAnomaly(ds);
  const L = eclipticLongitude(M);
  const dec = declination(L);
  const Jnoon = solarTransitJ(ds, M, L);
  const w = hourAngle(deg2rad(elevationDeg), phi, dec);
  if (Number.isNaN(w)) return null; // midnattssol / polarnatt
  const Jset = solarTransitJ(approxTransit(w, lw, n), M, L);
  return fromJulian(rising ? Jnoon - (Jset - Jnoon) : Jset);
}

export function computeSunTimes(date: Date, pos: LatLng): SunTimes {
  const now = sunPosition(date, pos);
  const sunrise = timeAtElevation(date, pos, -0.833, true);
  const sunset = timeAtElevation(date, pos, -0.833, false);
  const goldenStart = timeAtElevation(date, pos, 6, false); // kveld: sol < 6°
  const goldenMorningEnd = timeAtElevation(date, pos, 6, true);
  const blueStart = timeAtElevation(date, pos, -4, false);

  const iso = (d: Date | null) => (d ? d.toISOString() : "");
  return {
    sunriseIso: iso(sunrise),
    sunsetIso: iso(sunset),
    goldenHourStartIso: iso(goldenStart),
    goldenHourEndIso: iso(goldenMorningEnd),
    blueHourStartIso: iso(blueStart),
    azimuthDeg: now.azimuthDeg,
    elevationDeg: now.elevationDeg,
  };
}

/**
 * Lyskvalitet for et spottepunkt: sammenlign solretning med
 * fotograferingsretning. Sol bak fotografen = frontlys (bra),
 * 90° = sidelys (utmerket for fly), mot = motlys.
 */
export function lightQualityForDirection(
  sunAzimuthDeg: number,
  sunElevationDeg: number,
  shootingDirectionDeg: number,
): { quality: "excellent" | "good" | "fair" | "poor"; label: string } {
  if (sunElevationDeg < -4) return { quality: "poor", label: "Mørkt" };
  const relative = Math.abs(
    (((sunAzimuthDeg - shootingDirectionDeg) % 360) + 540) % 360 - 180,
  ); // 180 = sol rett bak fotografen, 0 = rett imot
  if (relative < 45) return { quality: "poor", label: "Motlys" };
  if (relative < 100) return { quality: "excellent", label: "Sidelys" };
  if (relative < 145) return { quality: "good", label: "Skrått frontlys" };
  return { quality: "good", label: "Frontlys" };
}

File: ipad/AeroSpotApp/AeroSpot/Services/SunService.swift
// SunService.swift — solposisjon og soltider (NOAA-formler).
// Porta 1:1 fra web-implementasjonens SunService.ts (13 tester grønne der).

import Foundation
import CoreLocation

enum SunService {
    private static let dayMs = 86_400_000.0
    private static let j1970 = 2440588.0
    private static let j2000 = 2451545.0
    private static let obliquity = rad(23.4397)
    private static let j0 = 0.0009

    private static func rad(_ deg: Double) -> Double { deg * .pi / 180 }

    private static func toJulian(_ date: Date) -> Double {
        date.timeIntervalSince1970 * 1000 / dayMs - 0.5 + j1970
    }
    private static func fromJulian(_ j: Double) -> Date {
        Date(timeIntervalSince1970: (j + 0.5 - j1970) * dayMs / 1000)
    }
    private static func toDays(_ date: Date) -> Double { toJulian(date) - j2000 }

    private static func solarMeanAnomaly(_ d: Double) -> Double {
        rad(357.5291 + 0.98560028 * d)
    }
    private static func eclipticLongitude(_ m: Double) -> Double {
        let c1 = rad(1.9148) * sin(m)
        let c2 = rad(0.02) * sin(2 * m)
        let c3 = rad(0.0003) * sin(3 * m)
        return m + c1 + c2 + c3 + rad(102.9372) + .pi
    }
    private static func declination(_ l: Double) -> Double {
        asin(sin(l) * sin(obliquity))
    }
    private static func rightAscension(_ l: Double) -> Double {
        atan2(sin(l) * cos(obliquity), cos(l))
    }
    private static func siderealTime(_ d: Double, _ lw: Double) -> Double {
        rad(280.16 + 360.9856235 * d) - lw
    }

    static func position(date: Date, coordinate: CLLocationCoordinate2D) -> (azimuthDeg: Double, elevationDeg: Double) {
        let lw = (-coordinate.longitude).degToRad
        let phi = coordinate.latitude.degToRad
        let d = toDays(date)
        let m = solarMeanAnomaly(d)
        let l = eclipticLongitude(m)
        let dec = declination(l)
        let ra = rightAscension(l)
        let h = siderealTime(d, lw) - ra

        let elevation = asin(sin(phi) * sin(dec) + cos(phi) * cos(dec) * cos(h))
        let azimuth = atan2(sin(h), cos(h) * sin(phi) - tan(dec) * cos(phi)) + .pi
        return ((azimuth.radToDeg + 360).truncatingRemainder(dividingBy: 360), elevation.radToDeg)
    }

    private static func julianCycle(_ d: Double, _ lw: Double) -> Double {
        (d - j0 - lw / (2 * .pi)).rounded()
    }
    private static func approxTransit(_ ht: Double, _ lw: Double, _ n: Double) -> Double {
        j0 + (ht + lw) / (2 * .pi) + n
    }
    private static func solarTransitJ(_ ds: Double, _ m: Double, _ l: Double) -> Double {
        j2000 + ds + 0.0053 * sin(m) - 0.0069 * sin(2 * l)
    }
    private static func hourAngle(_ h: Double, _ phi: Double, _ dec: Double) -> Double {
        acos((sin(h) - sin(phi) * sin(dec)) / (cos(phi) * cos(dec)))
    }

    /// Tidspunkt solen krysser gitt elevasjon; stigende (morgen) eller settende.
    static func timeAtElevation(
        date: Date,
        coordinate: CLLocationCoordinate2D,
        elevationDeg: Double,
        rising: Bool
    ) -> Date? {
        let lw = (-coordinate.longitude).degToRad
        let phi = coordinate.latitude.degToRad
        let d = toDays(date)
        let n = julianCycle(d, lw)
        let ds = approxTransit(0, lw, n)
        let m = solarMeanAnomaly(ds)
        let l = eclipticLongitude(m)
        let dec = declination(l)
        let jnoon = solarTransitJ(ds, m, l)
        let w = hourAngle(elevationDeg.degToRad, phi, dec)
        guard w.isFinite else { return nil } // midnattssol / polarnatt
        let jset = solarTransitJ(approxTransit(w, lw, n), m, l)
        return fromJulian(rising ? jnoon - (jset - jnoon) : jset)
    }

    static func times(date: Date, coordinate: CLLocationCoordinate2D) -> SunTimes {
        let now = position(date: date, coordinate: coordinate)
        return SunTimes(
            sunrise: timeAtElevation(date: date, coordinate: coordinate, elevationDeg: -0.833, rising: true),
            sunset: timeAtElevation(date: date, coordinate: coordinate, elevationDeg: -0.833, rising: false),
            goldenHourStart: timeAtElevation(date: date, coordinate: coordinate, elevationDeg: 6, rising: false),
            blueHourStart: timeAtElevation(date: date, coordinate: coordinate, elevationDeg: -4, rising: false),
            azimuthDeg: now.azimuthDeg,
            elevationDeg: now.elevationDeg
        )
    }

    enum LightQuality: String {
        case excellent, good, fair, poor
    }

    /// Lyskvalitet gitt solretning vs fotograferingsretning.
    static func lightQuality(
        sunAzimuthDeg: Double,
        sunElevationDeg: Double,
        shootingDirectionDeg: Double
    ) -> (quality: LightQuality, label: String) {
        if sunElevationDeg < -4 { return (.poor, "Mørkt") }
        let relative = abs(
            ((sunAzimuthDeg - shootingDirectionDeg).truncatingRemainder(dividingBy: 360) + 540)
                .truncatingRemainder(dividingBy: 360) - 180
        )
        if relative < 45 { return (.poor, "Motlys") }
        if relative < 100 { return (.excellent, "Sidelys") }
        if relative < 145 { return (.good, "Skrått frontlys") }
        return (.good, "Frontlys")
    }
}
